File: test/native/com/lib/flutter_pcm_sound/PcmClaimsTest.java
package com.lib.flutter_pcm_sound;

public final class PcmClaimsTest {
    private static void check(boolean value) { if (!value) throw new AssertionError(); }
    public static void main(String[] args) {
        PcmClaims c = new PcmClaims();
        long old = c.take(), current = c.take();
        check(!c.admit(old)); // a stale owner is refused, changing nothing
        check(c.admit(current));
        check(c.admit(current)); // admitting does not advance the claim
        check(c.admit(null)); // an unowned setup claims afresh...
        check(!c.admit(current)); // ...so it fences every older owner
        check(c.take() == current + 2);
    }
}

File: test/native/pcm_claims_test.c
#include "../../ios/Classes/PcmClaims.h"
#include <assert.h>
#include <stdio.h>

int main(void) {
    PcmClaims claims = {0};
    uint64_t old = PcmClaimsTake(&claims), current = PcmClaimsTake(&claims);
    assert(!PcmClaimsAdmit(&claims, true, old)); // a stale owner is refused, changing nothing
    assert(PcmClaimsAdmit(&claims, true, current));
    assert(PcmClaimsAdmit(&claims, true, current)); // admitting does not advance the claim
    assert(PcmClaimsAdmit(&claims, false, 0)); // an unowned setup claims afresh...
    assert(!PcmClaimsAdmit(&claims, true, current)); // ...so it fences every older owner
    assert(PcmClaimsTake(&claims) == current + 2);
    puts("PCM claim regressions passed");
}
